### PR TITLE
[WIP] generalized Newtonian fluid for incompressible flow solvers

### PR DESCRIPTION
**Pull request by richardschu**
This PR enables variable viscosity for the incompressible flow solvers and demonstrates the expected convergence rates via a manufactured solution to be found in `applications/incompressible_navier_stokes/navier_stokes_manufactured/`.
Contributions are:
1) Implement variable viscosity due to generalizted Newtonian fluids based on the implementation of turbulence models, merging the implementations, which both precompute the viscosity coefficients per integration point (see #353 ). This motivated including a number of Assertions in `parameters.cpp`. 
2) fixes a bug in the face evaluation of the coefficient
3) demonstration in `navier_stokes_manufactured` (everything as expected)
4) automated convergence table computation by writing and reading the error files, which were already produced, but now giving them identifiers and a `run_id`. This gives flexibility in computing the individual runs to be put into the convergence table and also does not break the interfaces. The location of the convergence study files was hard-coded to `output/` already before.
Note that I used the new formatting already, will rebase once #364 is merged, such that the changes are easier compared!

### CODE REVIEW

File: applications/incompressible_navier_stokes/navier_stokes_manufactured/input.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{
```
Comment (richardschu) [inline applications/incompressible_navier_stokes/navier_stokes_manufactured/input.json:1]: this settings demonstrates the expected convergence in space

File: include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.h [hunk 2]
```diff
@@ -68,17 +73,24 @@ class NonlinearOperatorCoupled
...
    */
   void
-  evaluate_residual(BlockVectorType & dst, BlockVectorType const & src) const
+  evaluate_residual(BlockVectorType & dst, BlockVectorType const & src)
   {
+    // update implicitly coupled viscosity
+    if(nonlinear_viscous_term_treated_implicitly == true)
+    {
+      pde_operator->update_viscosity(src.block(0));
+    }
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.h:82]: This should happen inside `evaluate_nonlinear_residual()`. `NonlinearOperatorCoupled` is a pure "linear algebra" wrapper without any physics implemented.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.h [hunk 3]
```diff
@@ -68,17 +73,24 @@ class NonlinearOperatorCoupled
...
   }
 
 private:
-  PDEOperator const * pde_operator;
+  PDEOperator * pde_operator;
 
   VectorType const * rhs_vector;
   double             time;
   double             scaling_factor_mass;
+  bool               nonlinear_viscous_term_treated_implicitly;
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/operator_coupled.h:93]: should not appear here in my opinion.

File: include/exadg/incompressible_navier_stokes/solver.h [hunk 4]
```diff
@@ -87,6 +95,126 @@ run(std::string const & input_file,
     driver->print_performance_results(timer.wall_time());
 }
 
+void
+evaluate_convergence_study(SpatialResolutionParameters const &  spatial,
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/solver.h:99]: I think this functionality should be realized as a separate PR. The present PR tends to be way too large.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/operator_pressure_correction.h [hunk 5]
```diff
@@ -68,15 +73,23 @@ class NonlinearMomentumOperator
   void
   evaluate_residual(VectorType & dst, VectorType const & src)
   {
+    // update implicitly coupled viscosity
+    if(nonlinear_viscous_term_treated_implicitly == true)
+    {
+      pde_operator->update_viscosity(src);
+    }
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/operator_pressure_correction.h:80]: Same as above.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/operator_pressure_correction.h [hunk 6]
```diff
@@ -68,15 +73,23 @@ class NonlinearMomentumOperator
...
 
 private:
-  PDEOperator const * pde_operator;
+  PDEOperator * pde_operator;
+  //  PDEOperator const * pde_operator;
 
   VectorType const * rhs_vector;
   double             time;
   double             scaling_factor_mass;
+  bool               nonlinear_viscous_term_treated_implicitly;
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/operator_pressure_correction.h:92]: same as above for coupled solver.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp [hunk 7]
```diff
@@ -860,26 +913,27 @@ SpatialOperatorBase<dim, Number>::get_constraint_u() const
...
-{
-  return param.viscosity;
-}
+// template<int dim, typename Number>
+// double
+// SpatialOperatorBase<dim, Number>::get_viscosity() const
+//{
+//  pcout << "THIS RETURNS A CONSTANT VISCOSITY ##+ spatial_operator_base.cpp\n";
+//  return param.viscosity;
+//}
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp:922]: TODO

File: include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.h [hunk 8]
```diff
@@ -654,9 +655,23 @@ class SpatialOperatorBase : public dealii::Subscriptor
...
+  ViscosityModel<dim,
+                 Number,
+                 true /*use_turbulence_model*/,
+                 true /*use_generalized_newtonian_model*/>
+    generalized_newtonian_turbulent_viscosity_model;
+  ViscosityModel<dim,
+                 Number,
+                 false /*use_turbulence_model*/,
+                 true /*use_generalized_newtonian_model*/>
+    generalized_newtonian_viscosity_model;
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.h:674]: This does not look like a modular design that can be extended in the future. This is something I would like to discuss in general. Maybe an issue is needed for this.

File: include/exadg/incompressible_navier_stokes/spatial_discretization/viscosity_model.cpp [hunk 9]
```diff
@@ -0,0 +1,838 @@
...
+
+      if(use_turbulence_model == true)
+      {
+        add_turbulent_viscosity(viscosity /*might use generalized Newtonian viscosity*/,
+                                filter_width,
+                                velocity_gradient,
+                                symmetric_velocity_gradient,
+                                shear_rate_squared,
+                                turbulence_model_data.constant);
+      }
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/spatial_discretization/viscosity_model.cpp:129]: I currently believe we should think about a more modular design where different models (living in `IncNS::SpatialOperatorBase`, e.g. `TurbulenceModel`, `GeneralizedNewtonian...`) may accumulate/assemble into the same `VariableCoefficients` field. This implies slightly more computational work. However, I otherwise have the concern that things get coupled too strongly in the code, making changes and maintenance more and more challenging.

File: include/exadg/incompressible_navier_stokes/user_interface/enum_types.cpp [hunk 10]
```diff
@@ -59,6 +59,19 @@ enum_to_string(ProblemType const enum_type)
   return string_type;
 }
 
+void
+string_to_enum(ProblemType & enum_type, std::string const string_type)
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/user_interface/enum_types.cpp:63]: I see these changes independent of the present PR and hence they should not be included here.

File: include/exadg/incompressible_navier_stokes/user_interface/parameters.cpp [hunk 11]
```diff
@@ -536,9 +548,96 @@ Parameters::check(dealii::ConditionalOStream const & pcout) const
...
+    if(generalized_newtonian_model == GeneralizedNewtonianModel::PowerLaw)
+    {
+      AssertThrow(std::abs(viscosity) < 1e-20,
+                  dealii::ExcMessage(
+                    "viscosity (= generalized_newtonian_kinematic_viscosity_lower_limit) == 0 "
+                    "required for GeneralizedNewtonianModel::PowerLaw."));
+      AssertThrow(std::abs(generalized_newtonian_kappa) < 1e-20,
+                  dealii::ExcMessage("generalized_newtonian_kappa == 0 required for "
+                                     "required for GeneralizedNewtonianModel::PowerLaw."));
+    }
```
Comment (nfehn) [inline include/exadg/incompressible_navier_stokes/user_interface/parameters.cpp:640]: I think these checks should not appear in `IncNS::Parameters`. We should follow a more modular design with certain classes caring about consistency of parameters for a particular functionality. This means that we should introduce a separate data struct with all parameters for generalized Newtonian, similarly for the already existing turbulence model stuff, etc.

File: include/exadg/postprocessor/error_calculation.cpp [hunk 12]
```diff
@@ -242,6 +250,64 @@ ErrorCalculator<dim, Number>::do_evaluate(VectorType const & solution_vector, do
   }
 }
 
+template<int dim, typename Number>
+std::string
+ErrorCalculator<dim, Number>::filename_from_filename_base(std::string const & directory,
```
Comment (nfehn) [inline include/exadg/postprocessor/error_calculation.cpp:255]: should be a separate PR.

**Comment by nfehn** [comment]
This pull request currently mixes too many different aspects and should be split up into several PRs in opinion. The most fundamental change appears to be the introduction of nonlinear viscosity models and different models for the viscosity. I think the design choices related to this should be at the heart of this PR, but there are another approx. 2000+ lines of code in the present PR that are not directly related to this.

**Comment by richardschu** [comment]
> This pull request currently mixes too many different aspects and should be split up into several PRs in opinion. The most fundamental change appears to be the introduction of nonlinear viscosity models and different models for the viscosity. I think the design choices related to this should be at the heart of this PR, but there are another approx. 2000+ lines of code in the present PR that are not directly related to this.

AGREED! Sorry for that and thanks for checking it out. What about first, you(?) merge #364 , then I can work on the PRs you commented earlier this week without triggering merge conflicts later and split this one up into 3, being:
1) automated convergence study (do we even want that? that was just an idea?)
2) string_to_enum or enum_to_string for easier parameter passing
3) generalized Newtonian fluid models + the testcase
What do you think?

**Comment by nfehn** [comment]
Point 1 sounds indeed reasonable/interesting in principle. In particular, I see a major benefit in such a functionality if we realize this as a utility functionality in ExaDG to be used by different modules.

Point 2 is (with the changes shown here) to be discussed in a broader context in my opinion. You suggested certain extensions in this PR, but many parameters are still not addressed. The problem that I see is that if we add certain parameters just for the module IncNS, it becomes more and more difficult to see for users which parameters can be set in an input .json file and which not. In any case, I suggest to first discuss this aspect before starting to implement this, in order to avoid that someone is disappointed because we have different opinions on this topic and cannot decide for a certain solution.
In light of the new generalized Newtonian functionality, we should first discuss how to structure our parameters in a hierarchical way so that the overall code remains modular and easy to maintain in case that more and more features are added. In the IncNS module, we typically pass `Parameters` to multiple classes, even though a certain class only needs a certain subset of the parameters. Reading parameters from an input .json file is an independent discussion in my opinion.

Point 3 appears to be most relevant in my opinion, for ExaDG and (probably) also for your own progress.

**Comment by btemuer** [comment]
I just want to post a reminder here regarding the variable viscosity/integration rule discussion in #353 where we found a preliminary solution with an assertion for standard quadrature. The assertion I put in there just checks against `use_turbulence_model` which has been the only case of variable viscosity until now. Apparently, this PR will change that, so we shouldn´t forget to adapt the assertion or find a better solution before this is merged.

**Comment by btemuer** [comment]
> In light of the new generalized Newtonian functionality, we should first discuss how to structure our parameters in a hierarchical way so that the overall code remains modular and easy to maintain in case that more and more features are added.

I want to draw attention to the hierarchic parameter structure that I had independently implemented for the Darcy module in my fork. You can find it [here]( 
https://github.com/btemuer/exadg/blob/a3e97704d44b4ff6bf937a8a3f2ae22003f2b8fe/include/exadg/darcy/user_interface/parameters.h). It just uses simple struct compositions. Maybe it can be an inspiration.
